### PR TITLE
feat: refactor expose ping logic

### PR DESCRIPTION
**Pull request by ktaube**
- Now uses `websockets` re-connection mechanisms
- Refactored payload handling to separate functions
- Workaround for `websockets` logger to not print stack trace when re-connecting

## Preview

https://github.com/robocorp/robocorp/assets/1479451/bebb9060-8c92-4138-a274-4eb1402c3ce6

### CODE REVIEW

File: action_server/src/robocorp/action_server/_server_expose.py [hunk 1]
```diff
@@ -124,180 +180,92 @@ async def expose_server(
...
+        headers = (
+            {
+                "x-session-id": session_payload.sessionId,
+                "x-session-secret": session_payload.sessionSecret,
+            }
+            if session_payload
+            else {}
+        )
+
+        async for ws in websockets.connect(
```
Comment (ktaube) [inline action_server/src/robocorp/action_server/_server_expose.py:177]: Uses websockets [own mechanisms](https://websockets.readthedocs.io/en/stable/faq/client.html#how-do-i-reconnect-when-the-connection-drops) to reconnect.

File: action_server/src/robocorp/action_server/_server_expose.py [hunk 2]
```diff
@@ -98,23 +118,74 @@ def forward_request(base_url: str, payload: BodyPayload) -> requests.Response:
...
 
-class ExposePingTimeoutError(Exception):
-    pass
+def handle_session_payload(
+    session_payload: SessionPayload, api_key: str | None, expose_url: str, datadir: str
+):
+    url = f"https://{session_payload.sessionId}.{expose_url}"
+    log.info(f"🌍 URL: {url}")
+    if api_key is not None:
+        log.info(
```
Comment (fabioz) [inline action_server/src/robocorp/action_server/_server_expose.py]: Question: shouldn't the api_key be checked here?

File: action_server/src/robocorp/action_server/_server_expose.py [hunk 3]
```diff
@@ -15,6 +16,25 @@
...
+
+console_handler = logging.StreamHandler()
+console_handler.setLevel(logging.DEBUG)
+
+formatter = NoExceptionFormatter("%(message)s")
+console_handler.setFormatter(formatter)
+
+
+websockets_logger = logging.getLogger("websockets")
+websockets_logger.addHandler(console_handler)
```
Comment (fabioz) [inline action_server/src/robocorp/action_server/_server_expose.py]: I think this shouldn't be done here... the logging is configured at `robocorp.action_server.cli._setup_stdout_logging` and `robocorp.action_server.cli._setup_logging`, so, this probably belongs there...

File: action_server/src/robocorp/action_server/_server_expose.py (hunk 2 above)
Comment (ktaube) [reply action_server/src/robocorp/action_server/_server_expose.py]: Not relevant after https://github.com/robocorp/robocorp/pull/161

File: action_server/src/robocorp/action_server/_server_expose.py (hunk 3 above)
Comment (ktaube) [reply action_server/src/robocorp/action_server/_server_expose.py]: `websockets` has its own re-connect logic, but it prints the failure logs with the stack trace making the output too verbose. This is just a workaround to stop it from printing the stack trace. See [websockets code](https://github.com/python-websockets/websockets/blob/35bc7dd8288445289134c335aae8af859862ccd1/src/websockets/legacy/client.py#L606) 

So I don't think this workaround should be moved to `robocorp.action_server.cli`.

Comment (fabioz) [reply action_server/src/robocorp/action_server/_server_expose.py]: The issue here is that the code is adding directly a logger and stopping propagation and the logger is redirected to stdout.

The action server actually adds 2 handlers, one for the stdout and another to a rotating file handler. The code here just adds one related to the stdout and stops the propagation for the root in the stdout, but then the rotating file handler doesn't get any notification.

This means that logs will always be shown in the console the way it's done here and not in the file and the log level here doesn't respect the log level that was configured in `_setup_stdout_logging`/`_setup_logging`, which is not what's expected (it's expected that the log level will respect the user setting and also it should redirect to both, stdout and the file).

Comment (ktaube) [reply action_server/src/robocorp/action_server/_server_expose.py]: Okay, any pointers how to

1. keep `websockets` re-connect logging
2. remove the stack trace

without my changes?

Comment (fabioz) [reply action_server/src/robocorp/action_server/_server_expose.py]: Maybe you could change the current formatters to format differently (i.e.: remove the exception info) based on the `module` and `funcName` available?  (in `_setup_stdout_logging`/`_setup_logging`)

i.e.: Create a new formatter which wraps an existing formatter and just implements `format` and removes the exception info if the record module/funcName match something you expect. That way you don't need to add a new handler, just change the existing formatters to be that formatter-wrapper.

Does it make sense?

**Comment by ktaube** [comment]
Had a call with Fabio and now I understand what changes I need to do for logging. Will do in separate PR.

**Review by fabioz — APPROVED** [review]

**Pull request merged**